Update nuke.c to auto-detect flash size

tyeth opened the pull request:
Fixes https://github.com/raspberrypi/pico-examples/issues/642


Original work by @Gadgetoid 🥇

tyeth committed:
Update nuke.c to auto-detect flash size

Fixes https://github.com/raspberrypi/pico-examples/issues/642


Thanks to @Gadgetoid for the work "Making A Thing" - Support [PiMoRoNi by buying stuff](https://shop.pimoroni.com/)!

tyeth committed:
Update nuke.c to auto-detect flash size

Fixes https://github.com/raspberrypi/pico-examples/issues/642


Thanks to @Gadgetoid for the work "Making A Thing" - Support [PiMoRoNi by buying stuff](https://shop.pimoroni.com/)!

lurch commented on flash/nuke/nuke.c:
```diff
@@ -27,12 +27,11 @@
+    uint8_t txbuf[4];
+    uint8_t rxbuf[4];
+    txbuf[0] = 0x9f;
+    flash_do_cmd(txbuf, rxbuf, 4);
```
Is this guaranteed to work with **every** variety of SPI Flash chip that might be used with RP2040 or RP2350?

lurch replied:
> I'm not an expert on this matter.

Me neither! The only reason I asked, was because I assumed that if it _was_ universally supported, then there'd already be a wrapper-function for it in pico-sdk? :thinking: 
But that's merely a hunch, I could be **entirely** wrong! :speak_no_evil:

tyeth replied:
This suggests for an FPGA situation with NOR flash that two possible command sets exist, but I've no idea if that first option is supported by RP2s.
https://electronics.stackexchange.com/a/673366

tyeth replied:
Okay, so digging deeper, yes the Serial Flash Discovery Protocol spec also mandates the backwards compatible support for 0x9f command, or at least as far as my limited research indicates.

kilograham replied:
cc @Wren6991 who i believe has said in the past that such a command is not universally supported

lurch replied:
> That said it’s kinda up to the vendor to verify flash_nuke works on their board and - if not - supply a customised alternative.

:joy: :rofl: :joy: :rofl: :joy: :rofl: 

If every vendor provided even _half_ of the support that Pimoroni did, the world would be a much better place :grinning:

tyeth replied:
Maybe there is a known case of returning wrong info, so some default would be wiser...
https://github.com/adafruit/circuitpython/blob/ffc42b9f0f4aa5e812de5302bc217e1dd5a0f8b6/ports/raspberrypi/supervisor/internal_flash.c#L81C35-L89C6

peterharperuk committed:
Add some validation of flash size

If flash_do_cmd returns a bad size for the flash or fails,
use a default flash size of 16MB

peterharperuk committed:
Add some validation of flash size

If flash_do_cmd returns a bad size for the flash or fails,
use a default flash size of 16MB

peterharperuk reviewed (changes requested):
-1 as this needs a fix to the sdk

will-v-pi reviewed (approved):
LGTM, and it still builds universal UF2s (https://github.com/will-v-pi/pico-sdk-uf2s/actions/runs/27211253008)

One thought from testing is it might be nice to turn on the LED when the erase starts, as it can take a long time on some chips? Picos are very fast, but my Pimoroni Pico Plus 2 takes forever.
